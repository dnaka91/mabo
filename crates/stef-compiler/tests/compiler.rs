use std::{
    fmt::{self, Display},
    fs,
    path::Path,
    sync::OnceLock,
};

use insta::{assert_snapshot, glob, with_settings};
use miette::{Diagnostic, MietteHandler, MietteHandlerOpts, NamedSource, Report, ReportHandler};
use stef_parser::Schema;

struct Wrapper<'a>(&'a MietteHandler, &'a dyn Diagnostic);

impl<'a> Wrapper<'a> {
    fn new(diagnostic: &'a dyn Diagnostic) -> Self {
        static HANDLER: OnceLock<MietteHandler> = OnceLock::new();

        Self(
            HANDLER.get_or_init(|| {
                MietteHandlerOpts::new()
                    .color(false)
                    .context_lines(3)
                    .force_graphical(true)
                    .terminal_links(false)
                    .width(120)
                    .build()
            }),
            diagnostic,
        )
    }
}

impl<'a> Display for Wrapper<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.debug(self.1, f)
    }
}

#[test]
fn validate_schema() {
    glob!("inputs/validate/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let schema = Schema::parse(input.as_str()).unwrap();
        let result = stef_compiler::validate_schema(&schema).unwrap_err();
        let report = Report::new(result).with_source_code(NamedSource::new(
            path.file_name().unwrap().to_string_lossy(),
            input.clone(),
        ));

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error_validate", Wrapper::new(&*report).to_string());
        });
    });
}

#[test]
fn resolve_schema_local() {
    glob!("inputs/resolve/local_*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let schema = Schema::parse(input.as_str()).unwrap();
        let result = stef_compiler::resolve_schemas(&[("test", &schema)]).unwrap_err();
        let report = Report::new(result).with_source_code(NamedSource::new(
            path.file_name().unwrap().to_string_lossy(),
            input.clone(),
        ));

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error_resolve", Wrapper::new(&*report).to_string());
        });
    });
}

#[test]
fn resolve_schema_import() {
    let input = fs::read_to_string(
        Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/inputs/resolve/datetime.stef"),
    )
    .unwrap();
    let datetime = Schema::parse(input.as_str()).unwrap();

    glob!("inputs/resolve/import_*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let schema = Schema::parse(input.as_str()).unwrap();
        let result = stef_compiler::resolve_schemas(&[("test", &schema), ("datetime", &datetime)])
            .unwrap_err();
        let report = Report::new(result).with_source_code(NamedSource::new(
            path.file_name().unwrap().to_string_lossy(),
            input.clone(),
        ));

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error_resolve", Wrapper::new(&*report).to_string());
        });
    });
}

#[test]
fn resolve_schema_remote() {
    let input = fs::read_to_string(
        Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/inputs/resolve/datetime.stef"),
    )
    .unwrap();
    let datetime = Schema::parse(input.as_str()).unwrap();

    glob!("inputs/resolve/remote_*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let schema = Schema::parse(input.as_str()).unwrap();
        let result = stef_compiler::resolve_schemas(&[("test", &schema), ("datetime", &datetime)])
            .unwrap_err();
        let report = Report::new(result).with_source_code(NamedSource::new(
            path.file_name().unwrap().to_string_lossy(),
            input.clone(),
        ));

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error_resolve", Wrapper::new(&*report).to_string());
        });
    });
}
