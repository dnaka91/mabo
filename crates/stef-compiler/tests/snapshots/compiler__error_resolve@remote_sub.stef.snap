---
source: crates/stef-compiler/tests/compiler.rs
description: "use datetime::timing;\n\nstruct Sample {\n    value: timing::Other @1,\n}"
input_file: crates/stef-compiler/tests/inputs/resolve/remote_sub.stef
---
  × type resolution failed
  ├─▶ failed resolving type in remote modules
  ╰─▶ definition ❬Y❭Other❬Y❭ not found in module ❬Y❭datetime::timing❬Y❭
   ╭─[remote_sub.stef:1:1]
 1 │ use datetime::timing;
 2 │ 
 3 │ struct Sample {
 4 │     value: timing::Other @1,
   ·                    ──┬──
   ·                      ╰── used here
 5 │ }
   ╰────

